File: 0890-lemonade-change/0890-lemonade-change.py
class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        if bills[0] > 5:
            return False
        five, ten = 0, 0

        for b in bills:
            if b == 5:
                five += 1
            elif b == 10:
                if five > 0:
                    five -= 1
                    ten += 1
                else:
                    return False
            else:
                if five > 0 and ten > 0:
                    five -= 1
                    ten -= 1
                elif five >= 3:
                    five -= 3
                else:
                    return False
        return True
        

File: 0890-lemonade-change/0890-lemonade-change.rs
impl Solution {
    pub fn lemonade_change(bills: Vec<i32>) -> bool {
        let (mut five, mut ten) = (0, 0);

        for &bill in &bills {
            match bill {
                5 => five+=1,
                10 => {
                    if five > 0 {
                        five -= 1;
                        ten += 1;
                    } else {
                        return false;
                    }
                },
                20 => {
                    if ten > 0 && five > 0 {
                        ten -= 1;
                        five -= 1;
                    } else if five >= 3 {
                        five -= 3;
                    } else {
                        return false;
                    }
                },
                _ => {}
            }
        }
        true
    }
}